src/relay/ir/hash.cc

```diff
@@ -0,0 +1,301 @@
+ public:
+  explicit RelayHashHandler() {}
+
+  /*!
+   * Compute hash of a node.
+   * \param ref The node to hash.
+   * \return the hash value.
+   */
+  size_t Hash(const NodeRef& ref) {
+    if (!ref.defined()) return ref.hash();
+
+    if (ref->derived_from<TypeNode>()) {
+      return TypeHash(Downcast<Type>(ref));
+    }
+    if (ref->derived_from<ExprNode>()) {
+      return ExprHash(Downcast<Expr>(ref));
+    }
+    return AttrHash(ref);
+  }
+
+  /*!
+   * Compute hash of the attributes.
+   * \param ref The attributes.
+   * \return the hash value
+   */
+  size_t AttrHash(const NodeRef& ref) {
+    if (!ref.defined()) { return ref.hash(); }
+    return AttrsHashHandler::Hash(ref);
+  }
+  /*!
+   * Compute hash of a Relay type.
+   * \param ref The type to hash.
+   * \param rhs The right hand operand.
+   * \return the hash value.
+   */
+  size_t TypeHash(const Type& type) {
+    if (!type.defined()) { return type.hash(); }
+    auto found = hash_map_.find(type);
+    if (found != hash_map_.end()) {
+      return found->second;
+    } else {
+      auto hash = this->VisitType(type);
+      hash_map_.insert({type, hash});
+      return hash;
+    }
+  }
+  /*!
+   * Compute the hash of an expression.
+   *
+   * \note We run graph structural equality checking when comparing two Exprs.
+   *   This means that AlphaEqualHandler can only be used once for each pair.
+   *   The equality checker checks data-flow equvalence of the Expr DAG.
+   *   This function also runs faster as it memomizes equal_map.
+   *
+   * \param expr The expression to hash.
+   * \return the hash value.
+   */
+  size_t ExprHash(const Expr& expr) {
+    if (!expr.defined()) return expr.hash();
+    auto found = hash_map_.find(expr);
+    if (found != hash_map_.end()) {
+      return found->second;
+    } else {
+      auto hash = this->VisitExpr(expr);
+      hash_map_.insert({expr, hash});
+      return hash;
+    }
+  }
+
+ protected:
+  /*!
+   * \brief Hash a DataType.
+   * \param dtype The dtype to hash.
+   * \return the hash value.
+   */
+  size_t DataTypeHash(const DataType& dtype) {
+    return ::tvm::AttrsHash()(dtype);
+  }
+
+  using AttrsHashHandler::VisitAttr_;
+  size_t VisitAttr_(const Variable* var) final {
+    auto it = hash_map_.find(GetRef<VarExpr>(var));
+    if (it != hash_map_.end()) {
+      return it->second;
+    }
+
+    return std::hash<std::string>()(var->name_hint);
+  }
+
+  // Type hashing
+  size_t VisitType_(const TensorTypeNode* tensor_type) final {
+    size_t hash = std::hash<std::string>()(tensor_type->_type_key);
+    hash = Combine(hash, DataTypeHash(tensor_type->dtype));
+    hash = Combine(hash, Hash(tensor_type->shape));
+    return hash;
+  }
+
+  size_t VisitType_(const IncompleteTypeNode* incomplete) final {
+    return GetRef<IncompleteType>(incomplete);
+  }
+
+  size_t VisitType_(const TypeVarNode* tyvar) final {
+    int index = BindVar(GetRef<TypeVar>(tyvar));
```

Comment:
Since both TypeVar/Var/Variable have two possible ways of occurrence:
- The point of declaration
- The point of visit

There are certain implications in this function that is not necessarily apparent from the code. Because Visit already checks the hashmap, this means the TypeVar itself is unbound. And we hash it by the free variable index, which implies graph equality.

Reply:
Consider leave a comment here about its implications